doc/design/diagnosis.md

```diff
@@ -0,0 +1,105 @@
+# SQLFlow Diagnostics
+
+For a typical SQLFlow method calling link:
+
+``` golang
+// pkg/sqlflowserver
+func Do(program string, sess *Session) string {
+  return workflow.Run(
+    couler.GenCode(
+      verifier.Verify(
+        resolver.ResolveProgram(parser.Parse(program)),
+        sess.DB)))
+}
+```
+
+Any of the above function calls can raise an error, it may be a syntax error on parsing, or verify fault on verifying prediction and training data schema, or the submitter program running error, etc.
+
+ This document issues an SQLFlow Error package `error` to improve the error message maintainability and readability and maintain various error messages in a rigorous approach.
+
+## Error Package
+
+In order to distinguish which package the error occurred, e.g., `parser`, `verify`, we represent an enumerated type `ErrType`. For each `ErrType`, there would be various types errors, for an example, verifying would fault on **attribute value type error** or **attribute does not exist**, we represent an enumerated type variable: `ErrCode`, The `ErrCode` can be the same among different `ErrType`.
+
+``` golang
+// ErrType represents the errors in a pacakge
+type ErrType int
+
+const (
+  pkgVerify ErrType iota
+  pkgParser
+  ...
+)
+
+func (t ErrType) String() string {
+  return []string{"verify", "parser", ...}[t]
+}
+
+
+// ErrCode represents a specific error type of a ErrType
+type ErrCode int
```

Comment:
Go prefers error types, not error code, maybe only keep error type?

Reply:
After talked with @typhoonzero , we have some conclusion as the following:

1. Move  `ErrCode`, uses can write the following code:
``` golang
var (
	ErrAttributeNotExists = sferr.TypeVerify.New("The attribute %s does not exists", "Please check WITH columns attribute.")
)

func Verify() error {
	// do some thing
	return ErrAttributeNotExists.Format("no_exists_attr")
}
```  

1. Should keep `SQLFlowError` struct, which can be used to return the traceback, the `traceback` feature would be designed in the future.

1. Need more discussion on `ErrorType`.